test/spec/test-stall-detection.js

```diff
@@ -0,0 +1,425 @@
+import { Upload } from 'tus-js-client'
+import { TestHttpStack, getBlob, waitableFunction } from './helpers/utils.js'
+
+// Custom HTTP stack that can simulate stalls
+class StallSimulatingHttpStack extends TestHttpStack {
+  constructor() {
+    super()
+    this.stallOnNextPatch = false
+  }
+
+  createRequest(method, url) {
+    const req = super.createRequest(method, url)
+
+    // If we want to simulate a stall on PATCH requests, override the send method
+    if (this.stallOnNextPatch && method === 'PATCH') {
```

Comment:
I think it would be useful to have a test that not only checks if requests without any progress events are detected as well, but also requests where the progress events stop in the middle of the request.